src/bootstrap/src/core/config/config.rs

```diff
@@ -2335,6 +2319,29 @@ impl Config {
                 None => None,
                 Some(commit) => {
                     self.download_ci_rustc(commit);
+
+                    let builder_config_path =
+                        self.out.join(self.build.triple).join("ci-rustc/builder-config");
+                    let ci_config_toml = Self::get_toml(&builder_config_path);
```

Comment:
We should make the `get_toml` function fallible, otherwise if the CI config fails to be parsed with the local `TomlConfig` (once a config change is made), this will abort bootstrap.

Reply:
Hmm. Imagine that we remove some field from `config.toml` in the future. Then when you run `x build` locally (or it runs on CI in a PR that removes the field), it will try to parse the old config from CI which still contains the field, and bootstrap will crash with `failed to parse TOML configuration 'config.toml': unknown field x`. Surely that's not what we want?

If the config fails to be parsed, we should probably just disable download-rustc.

Reply:
We don't remove fields directly to avoid causing breaking changes for other people. Instead, we deprecate them and keep them for a period (~3 months) before removing them completely. I'm not sure how to ensure programmatically that this won't happen, but disabling `download-rustc` due to an unsupported option (or another issue which could make things even worse) might not be ideal.

Regarding the policy of not removing options directly, I think it would be great to explain this on https://rustc-dev-guide.rust-lang.org. I will open a PR for that.

Reply:
It might not be just deletion, we can also e.g. rename values of existing options or do other backwards incompatible changes. It does not matter *when* we remove the option. Even if there's a 3 month period, the PR that removes the option will crash on this line if the old config.toml configured the option.

Of course, this won't probably happen very often, but we don't really have a way to test our backwards compatibility of config.toml, and it wouldn't be great if an unintended BC break resulted in CI being broken. That's why I suggested to disable `download-rustc` instead. 

My line of thinking is that `download-rustc` is an optimization, and if we find out in rare edge-cases that it's not possible to apply it (because we can't properly compare configs), then we can just temporarily disable it (in the given PR that makes a BC break in config.toml, for example). The disadvantage of this is of course that if the comparison logic starts failing because of something else, we might end up silently disabling `download-rustc` on CI without us knowing. To avoid this, it might indeed be a better trade-off if this crashes loudly. But we should add some comment here that if that happens, it was caused by a config.toml BC break, and we should explain what to do about it.

Reply:
Pushed https://github.com/rust-lang/rust/pull/129052/commits/af6eedc89e3d3e8bce01661b48c294d337d5539c, which disables `download-rustc` *only* if the CI rustc config has unknown fields. This means that any other errors will still be caught (which was a concern above). It's kind a tricky (like most of the bootstrap logic..), but I couldn't figured a better solution.

Reply:
Shame that we can't just access the error kind from `toml::de::Error`. But this is better than nothing. Thank you!